blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchContainerImpl.java

```diff
@@ -1,20 +1,111 @@
 import java.util.Iterator;
 import java.util.List;
 
 /**
  * @author Vivek Pandey
  */
 public class BranchContainerImpl extends BluePipelineContainer {
+
+
+    /**
+     * Orders branches with most recent activity first. Favorited branches are always at the top, also in recent
+     * order.
+     */
+    private static final Comparator<BluePipeline> BRANCH_COMPARITOR = new Comparator<BluePipeline>() {
+        @Override
+        public int compare(BluePipeline _pipeline1, BluePipeline _pipeline2) {
+            BranchImpl pipeline1 = (BranchImpl)_pipeline1;
+            BranchImpl pipeline2 = (BranchImpl)_pipeline2;
+
+            // If One pipeline isnt a favorite there is no need to go further.
+            if(pipeline1.isFavorite() && !pipeline2.isFavorite()) {
+                return -1;
+            }
+
+            if(!pipeline1.isFavorite() && pipeline2.isFavorite()) {
+                return 1;
+            }
+
+            BlueRun latestRun1 = pipeline1.getLatestRun();
+            BlueRun latestRun2 = pipeline2.getLatestRun();
+
+            // If a pipeline doesnt have a run yet, no need to go further.
+            if(latestRun1 != null && latestRun2 == null) {
+                return -1;
+            }
+
+            if(latestRun1 == null && latestRun2 != null) {
+                return 1;
+            }
+
+            //If neither have runs, lets just order by name.
+            if(latestRun1 == null && latestRun2 == null) {
+                return pipeline1.getName().compareTo(pipeline2.getName());
+            }
+
+            // If one run hasnt finished yet, then lets order by that.
+            if(latestRun1.getEndTime() != null && latestRun2.getEndTime() == null) {
+                return 1;
+            }
+
+            if(latestRun1.getEndTime() == null && latestRun2.getEndTime() != null) {
+                return -1;
+            }
+
+            // If both jobs have ended, lets order by the one that ended last.
+            if(latestRun1.getEndTime() != null && latestRun2.getEndTime() != null) {
+                if(latestRun1.getEndTime().getTime() > latestRun2.getEndTime().getTime()) {
+                    return -1;
+                }
+
+                if(latestRun1.getEndTime().getTime() < latestRun2.getEndTime().getTime()) {
+                    return 1;
+                }
+
+                return pipeline1.getName().compareTo(pipeline2.getName());
+            }
+
+            //If both jobs have not eneded yet, we need to order by start time.
+            if(latestRun1.getStartTime() != null && latestRun2.getStartTime() == null) {
+                return 1;
+            }
+
+            if(latestRun1.getStartTime() == null && latestRun2.getStartTime() != null) {
+                return -1;
+            }
+
+            if(latestRun1.getStartTime() != null && latestRun2.getStartTime() != null) {
+                if(latestRun1.getStartTime().getTime() > latestRun2.getStartTime().getTime()) {
```

Comment:
Pull stuff like `latestRun1.getStartTime().getTime()` into locals so this reads better.